javascript/extractor/lib/typescript/src/type_table.ts

```diff
@@ -421,17 +421,32 @@ export class TypeTable {
     return id;
   }
 
+  // type -> [id (not unfolded), id (unfolded)]
+  private idCache = new WeakMap<ts.Type, [number | undefined, number | undefined]>();
+
   /**
    * Gets the canonical ID for the given type, generating a fresh ID if necessary.
    *
    * Returns `null` if we do not support extraction of this type.
    */
   public getId(type: ts.Type, unfoldAlias: boolean): number | null {
+    const cached = this.idCache.get(type);
+    if (typeof cached !== "undefined") {
+        let value = cached[unfoldAlias ? 1 : 0];
+        if (typeof value !== "undefined") return value;
+    }
+    const setAndReturn = (v: number | null) => {
+        let cached = this.idCache.get(type) || [undefined, undefined];
+        cached[unfoldAlias ? 1 : 0] = v;
+        this.idCache.set(type, cached);
+        return v;
+    }
+    
     if (this.typeRecursionDepth > 100) {
       // Ignore infinitely nested anonymous types, such as `{x: {x: {x: ... }}}`.
       // Such a type can't be written directly with TypeScript syntax (as it would need to be named),
       // but it can occur rarely as a result of type inference.
-      return null;
+      return setAndReturn(null);
```

Comment:
This is technically incorrect, as a type might be seen at depth 101 and then we cache the fact that it can't be extracted. Then later the type is seen at a lower depth and could be extracted, but then we immediately give up because of the cached failure. It causes the extraction result to depend on the other in which we visit the types.

I remember trying to cache the result of `getId` many times and always rediscovering the reason why it doesn't work. Could you leave a comment here explaining the above?

With that said, we really don't get much milage out of these types compared to how expensive they are to extract, and I'd like to move away from extracting types (but still augmenting the call graph). So I'm OK with letting it slide in order to speed up extraction, and we should start talking about disabling type-extraction altogether.

Reply:
> With that said, we really don't get much milage out of these types compared to how expensive they are to extract, and I'd like to move away from extracting types (but still augmenting the call graph). So I'm OK with letting it slide in order to speed up extraction, but we should start talking about disabling type-extraction altogether.

Yeah, we don't get much from types now that we focus on security.  

Our recognition of HTML elements is improved a bunch though (for client-side XSS).  
So we would loose some results. 

Is it something we should ask product about?